_posts/2018-05-02-load-balanced-smart-proxies-with-haproxy.md

```diff
@@ -0,0 +1,138 @@
+date: 2018-05-02 12:42:45
+author: Christopher Duryee, Evgeni Golov
+tags:
+- foreman
+---
+
+In production Foreman and Katello setups you might like to load balance the work of Smart Proxies across a cluster. This post describes how to do this, utilizing Foreman 1.17, Katello 3.6 and [HAProxy](https://www.haproxy.org/).
+
+<!--more-->
+
+The approach described here is not as evolved as [@sean797](https://community.theforeman.org/u/sean797)'s proposal for [HA Smart Proxies](https://community.theforeman.org/t/user-survey-supporting-ha-smart-proxies/8832), but it is possible to implement this approach with today's codebase.
+
+The idea is that you can just take a load-balancer like HAProxy and configure it in front of a number of Smart Proxies without much changes to the clients.
+
+| ![load-balanced smart proxies](/static/images/blog_images/2018-05-02-loadbalanced_proxy_diag.png) |
+|:--:|
+| *Load-balanced smart proxies. Not all ports are shown in diagram.* |
+
+## Caveats
+
+### PuppetCA
+
+Puppetserver does not allow certificate signing to be load-balanced. The puppetserver keeps certificate information like the serial number counter and CRL on the filesystem, and if multiple writers use the same data it can result in corruption. To work around this, we pick one smart proxy (typically the first one) to be the signing server. We then tell clients to send CA requests to port 8141 on the load balancer, which gets redirected to 8140 on the smart proxy with signing capabilities.
+
+### Yum metadata
+
+Each smart proxy generates its own yum metadata for RPM repositories. This means that if a request to `https://smartproxy.example.test/repo/repodata/repomd.xml` goes to the first smart proxy, and the next request to fetch `repodata/<CHECKSUM>-primary.xml.gz` goes to the second smart proxy, the second request will 404. To work around this, we advise that port 443 should be sticky-sessioned.
+
+## Requirements
+* an existing working Foreman/Katello installation (we don't need to do any changes here, setup is the normal installation)
+* two or more machines for the Smart Proxy cluster
+* a load balancer - you can use a machine with HAProxy if you don't already have a load balancer
+* a common name in the DNS that points to the load balancer (we'll be using `smartproxy.example.test`)
+
+## Installing the first Smart Proxy
+### Generating Katello certificates
+When calling `foreman-proxy-certs-generate`, we have to pass `--foreman-proxy-cname smartproxy.example.test` to add the load-balanced name as an accepted `subjectAltName` to the list of names in the generated certificates. This allows clients to connect to each member of the smartproxy cluster as `smartproxy.example.test`
+
+### Preparing `custom-hiera.yaml`
+We need to configure Pulp to serve redirects in the streamer code to the name of the load-balancer, but this option is not exposed in the installer. Thus edit `/etc/foreman-installer/custom-hiera.yaml` and add the following line at the end:
```

Comment:
As one who is not so much into the details of Pulp, I stumble across the term "streamer code". Maybe it would be helpful to eloborate on why that's necessary or point to appropriate docs.